DIPs/1NNN-ADR.md

```diff
@@ -0,0 +1,479 @@
+
+```
+void foo(string s) {}
+
+foo(i"test $(4)"); // a CT error due to type mismatch. The compiler should suggest functions (see below) to remedy the user's problem.
+```
+
+### Concatenations
+
+In order to facilitate convenient formatting of long strings, if a `StringLiteral` follows an `InterpolatedString`,
+it is appended to the `InterpolatedString` in the parsing pass. (Not the lexer pass.)
+A `StringPostfix` may not appear on any of them.
+Such concatenations enable the various string literal types to be used as `InterpolatedStrings`,
+such as:
+
+```
+i""q{apples and $("bananas")}
+```
+yielding a tuple expression:
+```
+_d_interpolated_string!("apples and ", _d_interpolated_format_spec(null)), ("bananas")
+```
+
+### Library additions
+
+While the compiler only lowers to library calls and thus does not need to know about any of this, user functionality relies on a small library implementation.
+
+One suggested implementation would be:
+
+```
+struct _d_interpolated_string(Parts...) {
+        static:
+
+        private bool hasAllSpecs() {
+            foreach(part; Parts)
+               static if(is(typeof(part) == _d_interpolated_format_spec))
+                  if(part.spec is null)
+                       return false;
+            return true;
+        }
+
+        private string toFormatStringImpl(string defaultSpec) {
+           if(__ctfe) {
+                string ret;
+                foreach(part; Parts)
+                        static if(is(typeof(part) == _d_interpolated_format_spec)) {
+                                if(part.spec is null)
+                                        ret ~= defaultSpec;
+                                else
+                                        ret ~= part.spec;
+                        } else static if(is(typeof(part) : const(char)[]))
+                                ret ~= part;
+                        else static assert(0);
+                return ret;
+           } else assert(0);
+        }
+
+        private immutable(char*) toFormatStringzImpl() {
+                return toFormatString!(null).ptr;
+        }
+
+	// we may also add a lambda for escaping chars to this
+        public template toFormatString(string defaultSpec) {
+                enum toFormatString = toFormatStringImpl(defaultSpec);
+        }
+
+        static if(hasAllSpecs()) {
+                // alias this to a string, even a string literal, loses
+                // the implicit conversion to const(char)* needed by
+                // printf. so gotta use a wrapper that explicitly returns
+                // that :(
+		//
+		// A future compiler enhancement could allow alias this to
```

Comment:
How come future and not right in? This would allow:
```
string s = i"";
```

Yes?

Reply:
no, that wouldn't allow it. If the alias this thing was there, it would allow implicit conversions to a `writef` / `format` style argument list, so they wouldn't have to explicitly overload on it.

I actually think that would be a BAD thing, so I do NOT think we should do it, even if we can. I'm just trying to point out that it is possible in the future if other people disagree.

The `string s = i"";` thing will never work with this proposal, a tuple cannot be converted to a string without either a function (which is what `.idup` does in the sample implementation) or total compiler magic that I know a lot of people would object to.